### ISSUE TITLE
Learning?

### ISSUE THREAD

**Issue by DeepZono** [junyanz/pytorch-CycleGAN-and-pix2pix#904]
Hello, 

Thank you for providing the code. I am using a custom dataset and generator, however my approach is based on pix2pix. I have noticed that the loss_D_fake and loss _D_real never oscillate after the first epoch! The average loss for these is around 0.25 for 21,000 images. According to [soumith point 10](https://github.com/soumith/ganhacks), things are working when D has low variance and goes down over time. My discriminator loss plots show no variance at all after 2 epochs. 

Similarly, the generator loss "MSE loss" is also stuck near 0.25 averaged over 21,000 images. However, the L1 loss is steadily decreasing, = 1.6 after 1 epoch and 1.27 after 18th epoch. 

1) Visually, the result seems fine, but I am concerned that the generator converges too quickly to a non-optimum result and is able to easily fool the discriminator, any thoughts on this?

2) According to [goodfellow](https://arxiv.org/pdf/1701.00160.pdf), the Nash equilibrium is around D(x) = 0.5 for all x, seeing how mine is stuck at 0.25 when you combine Discriminator loss, what could be possibly done to alleviate the issue? I also tried to make the discriminator stronger by increasing LR and increasing the number of layers but the loss was still near 0.25

3) Also, since you have a "n_layer" implementation of your discriminator, how do you usually choose the layers? For example, I have 6 layers in my generator, what would be the ideal choice for N_layer_discriminator? 

Training Details: 
LR = 0.0001 or 0.00005
Learning Rate Decay = None
Dropout: None
No LeakyReLU's in Generator
BatchNorms = None in Generator
Loss: lsgan
BatchSize = 1 

### **Loss_D_Fake**
![image](https://user-images.githubusercontent.com/44021337/72986719-22476500-3de9-11ea-885b-f738cf2b334a.png)
**Loss_D_Real**
![image](https://user-images.githubusercontent.com/44021337/72986748-2ffcea80-3de9-11ea-8dd8-70ede9592414.png)
**Loss_G_L1**
![image](https://user-images.githubusercontent.com/44021337/72986813-502ca980-3de9-11ea-8f45-a60aebe516ec.png)
**Loss_G_GAN**
![image](https://user-images.githubusercontent.com/44021337/72986831-5ae73e80-3de9-11ea-8e87-febc61c77802.png)

**Comment by fa9r** [comment]
Hey, I think there is a bug in your code, because:
- the GAN losses don't osciliate.
- D and G are supposed to be adversarials, but both D and G GAN losses decrease at the same time.
- all GAN losses stay at around 0.25, which is what you would expect for untrained networks in LSGAN (for D(x)=0.5, we get L=0.5**2=0.25). I.e. neither D or G seems to have learned from the GAN losses.

Currently, your network is probably just optimizing for L1 loss and somehow ignoring the GAN losses, which can still lead to visually acceptable results (it's probably just a bit blurry and not very colorful right?). 

I'd recommend to turn off the L1 loss and then train again. If your GAN losses are working, your network should still learn something. If not, it will just generate random noise.

**Comment by junyanz** [comment]
I recommend that you use the default code, before making modifications. For the generator, you probably should use some form of ReLU and normalization layer.
